### PR TITLE
Revit_Engine: fix the copy of the DiffingConfig properties

### LINKED ISSUE #1333: Revit_Engine: RevitDiffing `CustomObjectDifferencesComparers` can sometimes hold duplicate values, resulting in duplicated differences in some edge cases
#### Description:

In `RevitDiffing()`, we create an instance of `DiffingConfig` if null and store the `RevitComparisonConfig` in there. We also allow the user to specify a `DiffingConfig`, which is then cloned in order to allow the downstream modifications, if required.

Due to how the copy of the DiffingConfig is done in the RevitDiffing, it can happen that we duplicate the `CustomObjectDifferencesComparers`:
https://github.com/BHoM/Revit_Toolkit/blob/1e07df9e9c437a987dc2b6e04695f3ecd9f22dd3/Revit_Engine/Compute/RevitDiffing.cs#L155-L159

This is because we should replace the line:
```cs
diffConfigClone.CustomObjectDifferencesComparers.AddRange(diffConfig.CustomObjectDifferencesComparers)
```

with:
```cs
diffConfigClone.CustomObjectDifferencesComparers.AddRange(diffConfig.CustomObjectDifferencesComparers.Except(diffConfigClone.CustomObjectDifferencesComparers))
```

in order to avoid re-adding existing entries.

This however generally does not creates problems, but in some edge cases some duplicated differences may be returned by the Diffing process if not corrected. For example, a recently added DiffingTest (see https://github.com/BHoM/DiffingTests_Prototypes/issues/10) highlights duplicate differences when `RevitParameterNumericTolerance` is used in the RevitComparisonConfig. This latter feature was never properly tested or used before within the Revit context, as it was simply re-exposing some backend functionality already available to the base diffing.

Additionally, we miss to copy some boolean properties of the original DiffConfig when doing the clone. For this reason, it is better to rework slightly the private method `GetRevitDiffingConfig()` to properly handle the copy and not only a first instantiation.

#### Steps to reproduce:

See see https://github.com/BHoM/DiffingTests_Prototypes/issues/10.

#### Expected behaviour:

No duplicated `CustomObjectDifferencesComparers` should be added by RevitDiffing; all properties should be correctly passed.

#### Test file(s):

See see https://github.com/BHoM/DiffingTests_Prototypes/issues/10.

### PR DESCRIPTION
**Pull request by alelom**
### Issues addressed by this PR

Closes #1333

Fixes some mistakes in the assignment of the properties to the DiffingConfig during the RevitDiffing.  
This removes some side-effects in edge cases like when using the `RevitComparisonConfig.ParameterNumericTolerances` feature.

### Test files

Run the tests in DiffingTests prototypes. All must pass. The relevant ones are the ones linked in https://github.com/BHoM/DiffingTests_Prototypes/issues/10: in the `RevitDiffingTests` project, look for  `PropertyNumericTolerances_Equal` and `PropertyNumericTolerances_Different` (in the test class `RevitObjectDifferencesTests`).

### Changelog

- Fixes bug that created side-effects in edge cases like when using the `RevitComparisonConfig.ParameterNumericTolerances` feature.

### CODE REVIEW

**Comment by FraserGreenroyd** [comment]
@alelom could we not include the DiffingTests as part of unit tests on this repo then if they're linked to Revit_Toolkit diffing operations?

**Comment by alelom** [comment]
> > > @alelom could we not include the DiffingTests as part of unit tests on this repo then if they're linked to Revit_Toolkit diffing operations?
> > 
> > 
> > Diffing Tests includes a dedicated revitdiffing project, which is why it is referenced as part of this pr. As per why they are there rather than here: I considered to move them in the the Revit toolkit (you probably remember me mentioning that). However, when you look at the diffing behaviour on the Revit side, there are places (most places) where it is incredibly helpful to step into the code that belongs to the base Diffing engine. Not to go off topic, anyways I think that we could find ways to separate them more, adding a Revit unit test project here, but I think that some "integration tests" would still be needed and would need to sit in an overarching toolkit like the diffingtests prototypes. I have been thinking about solutions to formalise this more, happy to workshop something together.
> 
> Makes sense - though I think longterm we should look to add DiffingTests to the BHoM_Engine at the very least, just so we can perform automated tests on PRs and a nightly basis via the bot to protect the diffing parts? But happy to discuss away from this PR - your reasoning for the above makes sense to me 👍

Definitely agree!

**Review by pawelbaran — CHANGES_REQUESTED** [review]
I have gone through the code and have left only one minor comment. However, 2 diffing tests do not pass on my build:

image

image

File: Revit_Engine/Compute/RevitDiffing.cs [hunk 1]
```diff
@@ -152,11 +153,6 @@ private static Diff Diffing(IEnumerable<object> pastObjects, IEnumerable<object>
                 if (diffConfig != null)
```
[the remark this answers is not in the document]
Comment (alelom) [reply Revit_Engine/Compute/RevitDiffing.cs:153]: No this is still needed, in case an user invokes the `RevitDiffing` method (the overload at line 114) and passes in a `null` diffingConfig.

[the remark this answers is not in the document]
Comment (pawelbaran) [reply Revit_Engine/Compute/RevitDiffing.cs:153]: But isn't `diffConfig` going to always be assigned after the addition you made in line 147?

image

[the remark this answers is not in the document]
Comment (alelom) [reply Revit_Engine/Compute/RevitDiffing.cs:153]: Actually yes you're right! :)

[the remark this answers is not in the document]
Comment (alelom) [reply Revit_Engine/Compute/RevitDiffing.cs]: Removed in https://github.com/BHoM/Revit_Toolkit/pull/1334/commits/18b371ea96246fd798d5cc5a3b436db53e6de349

**Comment by alelom** [comment]
> I have gone through the code and have left only one minor comment. However, 2 diffing tests do not pass on my build:

Just for the record, this was due to an issue in DiffingTests_Prototypes which has now been solved: https://github.com/BHoM/DiffingTests_Prototypes/issues/9

**Comment by pawelbaran** [comment]
@BHoMBot check core
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check null-handling
@BHoMBot check serialisation
@BHoMBot check versioning
@BHoMBot check installer

**Review by pawelbaran — APPROVED** [review]
All tests have passed, more testing done on Revit diffing with satisfying results.

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge
